### PR TITLE
allow branchprotector to protect branches matching a certain regex

### LINKED ISSUE #22657: allow branchprotector to protect branches matching a certain regex
**What would you like to be added**:

There only seems to be an exclude option - https://github.com/kubernetes/test-infra/blob/1bd2a72cbe81fde4066a0e96a3ac3d75027bc624/prow/config/branch_protection.go#L53-L55

It would be ideal if we could supply a 'include' regex

**Why is this needed**:

As we introduce automated tooling or in the case someone pushes branch by accident to a repo they become protected and it becomes cumbersome to delete those.

ie. ideally we want to only protected `main` and `release-*` branches - so our inclusion regex would be `release-.*|main`

### PR DESCRIPTION
**Pull request by mcbenjemaa**
Add Include policy setting for branchprotector to allow protect certain branches.
Include can be emtpy and it has no effect. on current config.

Fixes #22657

### CODE REVIEW

File: prow/cmd/branchprotector/protect.go [hunk 1]
```diff
@@ -337,7 +345,12 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		}
 		for _, b := range bs {
 			_, ok := repo.Branches[b.Name]
-			if !ok && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
+			if !ok && branchInclusions != nil && branchInclusions.MatchString(b.Name) {
+				branches[b.Name] = b
+			} else if !ok && branchInclusions != nil && !branchInclusions.MatchString(b.Name) {
+				logrus.Infof("%s/%s=%s: excluded", orgName, repoName, b.Name)
+				continue
+			} else if !ok && branchInclusions == nil && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
```
Comment (alvaroaleman) [inline prow/cmd/branchprotector/protect.go:353]: We know for sure at this point that `branchInclusions` is nil

```suggestion
			} else if !ok && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
```

File: prow/cmd/branchprotector/protect.go [hunk 2]
```diff
@@ -321,6 +321,14 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		return nil
 	}
 
+	var branchInclusions *regexp.Regexp
```
Comment (alvaroaleman) [inline prow/cmd/branchprotector/protect.go:324]: Please add tests for this new functionality

File: prow/config/branch_protection.go [hunk 3]
```diff
@@ -53,6 +53,9 @@ type Policy struct {
 	// Exclude specifies a set of regular expressions which identify branches
 	// that should be excluded from the protection policy
 	Exclude []string `json:"exclude,omitempty"`
+	// Include specifies a set of regular expressions which identify branches
+	// that should be included from the protection policy
+	Include []string `json:"include,omitempty"`
```
Comment (alvaroaleman) [inline prow/config/branch_protection.go:58]: Please document that this is mutually exclusive with `Exclude` and add a check into the config load that errors if both are set

File: prow/cmd/branchprotector/protect.go [hunk 4]
```diff
@@ -337,7 +345,12 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		}
 		for _, b := range bs {
 			_, ok := repo.Branches[b.Name]
-			if !ok && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
+			if !ok && branchInclusions != nil && branchInclusions.MatchString(b.Name) {
+				branches[b.Name] = b
+			} else if !ok && branchInclusions != nil && !branchInclusions.MatchString(b.Name) {
+				logrus.Infof("%s/%s=%s: excluded", orgName, repoName, b.Name)
```
Comment (alvaroaleman) [inline prow/cmd/branchprotector/protect.go:351]:
```suggestion
				logrus.Infof("%s/%s=%s: not included", orgName, repoName, b.Name)
```

File: prow/cmd/branchprotector/protect.go [hunk 5]
```diff
@@ -337,7 +345,12 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		}
 		for _, b := range bs {
 			_, ok := repo.Branches[b.Name]
-			if !ok && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
+			if !ok && branchInclusions != nil && branchInclusions.MatchString(b.Name) {
+				branches[b.Name] = b
+			} else if !ok && branchInclusions != nil && !branchInclusions.MatchString(b.Name) {
+				logrus.Infof("%s/%s=%s: not included", orgName, repoName, b.Name)
+				continue
+			} else if !ok && branchExclusions != nil && branchExclusions.MatchString(b.Name) {
```
Comment (alvaroaleman) [inline prow/cmd/branchprotector/protect.go:353]: Branchprotector respecting branchInclusions needs tests

File: prow/config/branch_protection.go [hunk 6]
```diff
@@ -53,6 +53,9 @@ type Policy struct {
 	// Exclude specifies a set of regular expressions which identify branches
 	// that should be excluded from the protection policy
 	Exclude []string `json:"exclude,omitempty"`
```
Comment (alvaroaleman) [inline prow/config/branch_protection.go:55]: Please also doc on `Exclude` that it is mutually exclusive with `Include`

File: prow/cmd/branchprotector/protect.go (hunk 5 above)
Comment (mcbenjemaa) [reply prow/cmd/branchprotector/protect.go:353]: I have added a test case in the `TestApply`

File: prow/config/branch_protection_test.go [hunk 7]
```diff
@@ -289,6 +289,18 @@ func TestApply(test *testing.T) {
 				Exclude: []string{"bar*", "foo*"},
 			},
 		},
+		{
+			name: "merge inclusion strings",
```
Comment (alvaroaleman) [inline prow/config/branch_protection_test.go:293]: @mcbenjemaa this test seems to fail

**Comment by mcbenjemaa** [comment]
Ready

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**